src/Consolonia.Core/Text/Fonts/AsciiArtTypeface.cs

```diff
@@ -0,0 +1,405 @@
+        /// </summary>
+        /// <param name="glyphs">Sequence of glyphs to process</param>
+        /// <returns>Array of advances for each glyph</returns>
+        public byte[] CalculateAdvancesWithKerning(ReadOnlySpan<AsciiArtGlyph> glyphs)
+        {
+            if (glyphs.Length == 0)
+                return Array.Empty<byte>();
+
+            byte[] advances = new byte[glyphs.Length];
+
+            for (int i = 1; i < glyphs.Length; i++)
+            {
+                AsciiArtGlyph leftGlyph = glyphs[i - 1];
+                AsciiArtGlyph rightGlyph = glyphs[i];
+
+                if (rightGlyph.Codepoint == ' ' ||
+                    rightGlyph.Codepoint == 0x2009 || // THIN SPACE
+                    rightGlyph.Codepoint == '\t')
+                {
+                    // space glyph, no kerning/smushing
+                    advances[i - 1] = leftGlyph.Width;
+                    continue;
+                }
+
+                // Calculate maximum overlap between glyphs
+                int maxOverlap = CalculateMaxOverlap(leftGlyph, rightGlyph);
+
+                // Advance is the width minus the overlap; clamp to avoid underflow
+                int width = leftGlyph.Width;
+                int overlap = Math.Min(maxOverlap, width);
+                int advance = Math.Max(0, width - overlap);
+                advances[i - 1] = (byte)advance;
+            }
+
+            // last glyph advance is its full width
+            advances[^1] = glyphs[^1].Width;
+            return advances;
+        }
+
+        /// <summary>
+        ///     Calculate the maximum overlap (kerning/smush amount) between two glyphs
+        /// </summary>
+        private int CalculateMaxOverlap(AsciiArtGlyph leftGlyph, AsciiArtGlyph rightGlyph)
+        {
+            int minMove = int.MaxValue;
+
+            // Check each line to find the minimum move (tightest constraint)
+            int lineCount = Math.Max(leftGlyph.Lines.Length, rightGlyph.Lines.Length);
+
+            for (int lineIdx = 0; lineIdx < lineCount; lineIdx++)
+            {
+                string leftLine = lineIdx < leftGlyph.Lines.Length ? leftGlyph.Lines[lineIdx] : "";
+                string rightLine = lineIdx < rightGlyph.Lines.Length ? rightGlyph.Lines[lineIdx] : "";
+
+                int leftEnd = lineIdx < leftGlyph.Ends.Length ? leftGlyph.Ends[lineIdx] : -1;
+                int rightStart = lineIdx < rightGlyph.Starts.Length ? rightGlyph.Starts[lineIdx] : rightLine.Length;
+
+                // Calculate SpaceAfter (spaces after last character in left glyph)
+                int spaceAfter = leftEnd < 0 ? leftLine.Length : leftLine.Length - leftEnd - 1;
+
+                // Calculate SpaceBefore (spaces before first character in right glyph)
+                int spaceBefore = rightStart >= rightLine.Length ? rightLine.Length : rightStart;
+
+                // Base move is just the sum of whitespace margins
```

Comment:
These comments just repeat the code

Reply:
removed some of the comments, I kept the ones that help me understand the code.